### ISSUE TITLE
Support Embedded Relationships

### ISSUE THREAD

**Issue by richmolj** [json-api/json-api#1089]
### Intent

A number of issues have tried to address this in the past (https://github.com/json-api/json-api/issues/202, https://github.com/json-api/json-api/issues/205, https://github.com/json-api/json-api/issues/753, https://github.com/json-api/json-api/issues/536). Most of those issues were closed in favor of https://github.com/json-api/json-api/issues/795, and most Discourse threads on this topic eventually point to #795.

However, #795's scope involves two main points:
- Creating/updating a single resource with its relations in one go
- Multiple operations in a single request (primarily to support transactions)

This issue's scope is **only the former**. The intent is to resolve a common issue without trying to solve every other related issue at the same time. In other words, let's support something like [accepts_nested_attributes_for](http://api.rubyonrails.org/classes/ActiveRecord/NestedAttributes/ClassMethods.html) in Rails.
### Scope

As compared to https://github.com/json-api/json-api/issues/795:

``` diff
- * It should allow mixed actions in a single request (e.g. creating + updating).
+ * It should allow creating/updating a resource and its relationships in a single request
* It should work for multiple types of resources (e.g. articles + tags) that would otherwise be accessed
at different endpoints.
* Every request should be transactional (i.e. all operations should succeed or fail together).
* Additive to the base spec (probably as an extension).
+ * Minimal footprint possible, favoring #795 for more advanced use cases
+ * Forwards-compatible with the larger scope of #795.

Here are a couple use-cases that test these requirements:

- * Create two articles, add one existing tag to both, create a new tag and add to both.
* Change the title of an article and update its tags: add one existing tag, create another,
and remove another (without replacing the whole tag set).
```
### Example

<img width="429" alt="screen shot 2016-09-01 at 9 29 31 am" src="https://cloud.githubusercontent.com/assets/55264/18484902/69880558-79b8-11e6-8a7f-111168783df0.png">

In this form, we need to:
- Add already-persisted tags (say, from an autocomplete)
- Add unpersisted tags (based on user input)
- Remove tags (when editing)
- Submit the post and its tags at the same time, so validations can run together.
### Proposal
- Embrace @beauby's [proposal for dealing with partial updates to collections](https://github.com/json-api/json-api/issues/1097). TL;DR: introduce `added-data`, `removed-data`, and `updated-data` keys for partial updates, keeping `data` for a full replacement of the collection.
- Allow resource objects inside the relationship `data` field, not just resource identifiers (for instance, send `attributes` in the relationship's `data` key).

This has been proposed in the past, most recently by @andruby (https://github.com/json-api/json-api/issues/795#issuecomment-159926461). In fact, a number of users _already assume this is part of the spec_ (https://github.com/json-api/json-api/issues/795#issuecomment-159926461, https://github.com/json-api/json-api/issues/795#issuecomment-159927027, https://github.com/json-api/json-api/issues/795#issuecomment-168732087, https://github.com/cerebris/jsonapi-resources/issues/424#issuecomment-174629986, https://github.com/json-api/json-api/issues/795#issuecomment-202810800).

As opposed to creating a post and assigning pre-existing tags (currently supported by the spec):

**Scenario A** Creation:

Create a post and its tags in a single request:

``` diff
{
  data: {
    type: 'posts',
    attributes: { title: 'my title' },
    relationships: {
      tags: {
        data: [
-         { id: 1, type: 'tags' },
+         { type: 'tags', attributes: { name: 'mytag' } },
        ]
      }
    }
  }
}
```

**Scenario B**: Update:

Update the post's title, create a new tag, update an existing tag, and remove an existing tag.

NB: `data` is still valid for replacing the entire collection.

``` diff
{
  data: {
    type: 'posts',
    attributes: { title: 'my title' },
    relationships: {
      tags: {
-        data: [
-          { id: 1, type: 'tags' },
-        ]
+        updated-data: [
+          { id: 1, type: 'tags', attributes: { name: 'updated' } }
+        ],
+        added-data: [
+          { type: 'tags', attributes: { name: 'anothertag' } }
+        ],
+        removed-data: [
+          { id: 1, type: 'tags' }
+        ]
        ]
      }
    }
  }
}
```

**Scenario C** Multi-level nesting:

Create a post with a tag, and the tag's creator, in a single request. The `creator` here could have attributes specified as well. This can go N levels deep:

``` diff
{
  data: {
    type: 'posts',
    attributes: { title: 'my title' },
    relationships: {
      tags: {
        data: [
          {
            id: 1,
            type: 'tags',
+           relationships: {
+             creator: {
+               data: { id: 1, type: 'users' }  
+             }
+           }
          }
        ]
      }
    }
  }
}
```

**Scenario D** Destruction:

This draws the distinction between disassociating a tag (see above) and deleting it altogether.

``` diff
{
  data: {
    type: 'posts',
    attributes: { title: 'my title' },
    relationships: {
      tags: {
+       deleted-data: [
+          { id: 1, type: 'tags' }
+       ]
      }
    }
  }
}
```
### Pros
- Minimal and intuitive. As noted above, many users assume this is already part of the spec.
- Solves a common use case.
- Does not require client-generated ids (temporary or otherwise)
### Cons
- Impossible to reference an embedded resource in more than one relationship. See @ethanresnick's example [here](https://github.com/json-api/json-api/issues/536#issuecomment-96635505).
  
  I'd like to highlight this part:
  
  > Problems of this sort only become more common when this strategy is used in conjunction with the bulk extension. For example, suppose that I want to create multiple articles at once and tag both of them with the same new tag...this is impossible. Or, on a job board, I want to add multiple job postings at once and link them with the same new employer.
  
  I agree these types of problems mostly come up when dealing with bulk operations. So I would suggest not solving this use case as part of this issue. If users have this issue, they can use a TBD operations extension currently in discussion in #795.
- There are some concerns the premise of nested resources breaks REST principles. Per the description of #795, it does look like a goal of JSON API to support nested resources, but this is a valid issue. For a discussion of this area please see @brainwipe's [original comment](https://github.com/json-api/json-api/issues/1089#issuecomment-248235594), then read down to [this comment closing the discussion](https://github.com/json-api/json-api/issues/1089#issuecomment-248648558). If you view the concept of any form of nested resources as inherently a Bad Idea, please downvote this issue (and lend your feedback to #795 as well).
### Alternatives
- Sideposting: It looks like [this proposal](https://github.com/json-api/json-api/issues/536#issuecomment-96635505) by @ethanresnick was reaching general consensus before it was closed in favor of #795. I can get on board with something like this as well, though I favor the smaller footprint and more intuitive approach defined above.
- Operations: We could punt on this and fold it into #795, which has been the current approach. I would like to avoid this since embedded resources is an incredibly common requirement, operations less so. Embedded resources can be solved without biting off the additional complexity and RPC-style semantics.
### Client-Side Ephemeral IDs

One problem with the above approach is getting a reference to the newly-created nested resource on the client. In the case of Ember, for instance, you'd be left with an in-memory record with `isNew` set to `true`, and no `id`.

There are two ways to solve this:
- **A** The server response must sideload any nested relationship sent in the request. This way clients can drop the records that were just sent to the server and replace them with the contents of `included`. I have been using this pattern with Rails and ember-data without problems.
- **B** Ephemeral client-side IDs. The client sends something like `temp-id: 'abc-123' }` as part of the record, the server responds with the same.

I believe **A** should be the default, and **B** should be an optional implementation left up to users to accommodate more advanced use cases. _However_ I have no problems if the community reached a consensus that **B** was a better option. I suggest avoiding 'sideposting' as it makes client-side temporary ids _required_, whereas I'd prefer them _optional_.

(_I will continue to update this proposal as consensus is reached in the comments. Last updated: September 21, 2016_)

**Comment by sandstrom** [comment]
First of all, thanks for writing this up. A solution to embedding objects would be _very_ useful!

Also, I think it's good to separate this discussion from the operations (#795) which would also be useful, but is probably better tackled separately from embedded which, hopefully, could have a narrow scope and get implemented faster (and will also be simpler to implement server-side and client-side).

Some thoughts:
- Would placing `_disassociate` / `_delete` inside `meta` (instead of `attributes`) be better?
- Instead of having both `_disassociate` and `_delete`, one could use only, say, `_delete` and let the server decide if it should disassociate or delete. Such behaviour is quite common in server-side ORMs, where `mother.children.pop` may both disassociate and delete. Not sure about this though.
- I think ephemeral ids (to associate client-side models with those returned, as discussed under _Additional Information_) would be useful. Perhaps they should be considered? (note that I'm not thinking of client-side generated IDs here, ephemeral ids would only live for a short duration, to correlate the nested objects in request/response; also they wouldn't be persisted on the server)
- It would be useful to clarify how partial collections are handled (B, D). E.g. difference between POST/PUT.
- I wouldn't mind if this feature would be restricted to require that all embedded objects are assumed to be included, since in practice this will mostly be used with small sets of embedded collections.

**Comment by richmolj** [comment]
> Would placing _disassociate / _delete inside meta (instead of attributes) be better?

Yes, I think this is a good idea 👍 . I have updated the original proposal.

> Instead of having both _disassociate and _delete, one could use only, say, _delete and let the server decide if it should disassociate or delete. Such behaviour is quite common in server-side ORMs, where mother.children.pop may both disassociate and delete. Not sure about this though.

I personally have use cases where I want the same endpoint to handle both cases. So the server would need to know which was requested.

> I think ephemeral ids (to associate client-side models with those returned, as discussed under Additional Information) would be useful. Perhaps they should be considered?

Not against it, though I would probably favor the sideposting proposal linked to in 'alternatives' at that point. If we're requiring client-side IDs and thus accepting their downsides, might as well get all the benefits.

> It would be useful to clarify how partial collections are handled (B, D). E.g. difference between POST/PUT

Could you expand a little more on this, please? The general premise is you never have to supply the full collection, everything is always partial as far as the spec is concerned. So if you passed a nested tag in an update, that would append to the tags array, not replace it.

> I wouldn't mind if this feature would be restricted to require that all embedded objects are assumed to be included, since in practice this will mostly be used with small sets of embedded collections.

Yup, that's my proposal as well and what my current monkey patches are doing 👍

**Comment by spieker** [comment]
Thank you for opening a new issue on this topic.

I've one addition to the creation of new records. When taking the first example, it' impossible to "guess" the ID of created relationship records out of the response.

Lets say we have a request with two new tags:

``` js
{
  data: {
    type: 'posts',
    attributes: { title: 'my title' },
    relationships: {
      tags: {
        data: [
          { id: 1, type: 'tags' },
          { type: 'tags', attributes: { name: 'mytag1' } },
          { type: 'tags', attributes: { name: 'mytag2' } },
        ]
      }
    }
  }
}
```

and a response with the newly created records and it's IDs

``` js
{
  data: {
    type: 'posts',
    attributes: { title: 'my title' },
    relationships: {
      tags: {
        data: [
          { id: 1, type: 'tags' },
          { id: 2, type: 'tags', attributes: { name: 'mytag1' } },
          { id: 3, type: 'tags', attributes: { name: 'mytag2' } },
        ]
      }
    }
  }
}
```

Now EmberJS for example is holding the tag records in the store of the application, when receiving the response from the server two new models with the IDs 2 and 3 are added to it instead of updating the existing once.

I'd suggest to add a temporary ID to the create request that the server must add to the response so the models can be identified and updated on client side.

Request

``` js
{
  data: {
    type: 'posts',
    attributes: { title: 'my title' },
    relationships: {
      tags: {
        data: [
          { id: 1, type: 'tags' },
          { type: 'tags', _id: '0a15de4b-de73-4a0a-8772-9ace1b39c9d3', attributes: { name: 'mytag1' } },
          { type: 'tags', _id: 'ad0b71bd-0d58-488e-80b9-0894cc50e004', attributes: { name: 'mytag2' } },
        ]
      }
    }
  }
}
```

Response:

``` js
{
  data: {
    type: 'posts',
    attributes: { title: 'my title' },
    relationships: {
      tags: {
        data: [
          { id: 1, type: 'tags' },
          { id: 2, _id: '0a15de4b-de73-4a0a-8772-9ace1b39c9d3', type: 'tags', attributes: { name: 'mytag1' } },
          { id: 3, _id: 'ad0b71bd-0d58-488e-80b9-0894cc50e004', type: 'tags', attributes: { name: 'mytag2' } },
        ]
      }
    }
  }
}
```

**Comment by richmolj** [comment]
> I've one addition to the creation of new records. When taking the first example, it' impossible to "guess" the ID of created relationship records out of the response.

@spieker this is in the "Client-Side Ephemeral IDs" section of the issue. If all nested relationships are sideloaded in the response, clients can drop whatever was sent and replace with `included`. In my ember app, for instance, I do this:

``` js
// This code finds any records that were just sent to the server and are
// still marked isNew, or still marked for deletion/disassociation
//
// belongsTo works out of the box, this is specific to hasMany
const resetRelations = function(record) {
  // __recordsJustSaved is set in my serializer
  Object.keys(record.get('__recordsJustSaved')).forEach((relationName) => {
    let relationRecords = record.get('__recordsJustSaved')[relationName];

    relationRecords.forEach((r) => {
      let shouldUnload = r.get('isNew') || r.get('markedForDestruction');
      if (shouldUnload) {
        r.unloadRecord();
      } else if (r.get('markedForDeletion')) {
        record.get(relationName).removeObject(r);
        r.set('markedForDeletion', false);
      }
    });
  });
  record.set('__recordsJustSaved', []);
};

save(options = {}) {
  let promise = this._super(...arguments);
  promise.then(resetRelations);
  return promise;
}
```

If we can get this merged into the spec itself I think we could put similar code into ember-data directly and not have to worry about it.

(_I have updated the original post with a better description of this issue_.)

**Comment by beauby** [comment]
As pointed during our conversation @richmolj, relying on the `meta` member of a resource identifier object would be problematic because implementers would then have to be careful about what they put in `meta` in order to avoid data being deleted accidentally. 
The way I understand it, `meta` should be considered as off-limit to the spec, it is purely the implementer's backyard.

Instead, I'd suggest, as outlined [in this blog post](http://lucashosseini.com/extending-jsonapi/) to have updated relationships contain either a `data` member, the semantic of which would be "replace with provided value", or one or both of `added_data`, `removed_data`, the semantic of which would be, well "add provided resources to the relationship" and "remove provided resources from the relationship", respectively. 

Note that those would not, per se, handle the creation/deletion of related resources, but simply the modification of a relationship. Creation/deletion of related resources could still be handled via embedding (for creation), and sideposting (possibly introducing toplevel `added_data` and `removed_data`), with the need of temporary IDs for sideposting in `added_data`.

**Comment by brainwipe** [comment]
I am concerned that embedded relationships is an over complication and leaks the server side domain into the client. The example you gave seem innocuous enough but when you implement this in a real system, I think you might start breaking principals of REST, resources, domain driven design or encapsulation.

I'll explain this using some questions regarding your example. 

Are tags a means to an end (filtering Posts) or are they a resource on its own? Would you be filtering Posts by tag? Does it make sense for a tag exist without a Post?

It seems to me that a tag is a child entity of Post, not its own resource. If you wanted a list of tags then that would seem like a reasonable `GET` but adding/removing tags means little without there being a Post for them to be attached to.

A resource is a bounded context (or encapsulated entity, or aggregate). As per REST, you should be able to create/update/delete without the need to affect any other bounded context. It's a resource that is managed and maintained by itself. That is where its simplicity and power comes from. It also means that some domains are not suitable for REST. I've had difficulty with graph networks, hierarchies and deep compositional models with REST.

If tag is not a resource in of itself then it belongs inside `attributes` in the standard spec and it's up to the server to break them up and manage them.

The example is a little too trivial to go too far with. Perhaps I need to see another example, which is a slightly more complex domain (with lots of attributes in the child entity) to be sure that this isn't huge complexity for the sake of not rethinking the server side domain model.

**Comment by richmolj** [comment]
> As pointed during our conversation @richmolj, relying on the meta member of a resource identifier object would be problematic because implementers would then have to be careful about what they put in meta in order to avoid data being deleted accidentally. 
> The way I understand it, meta should be considered as off-limit to the spec, it is purely the implementer's backyard.

Sold. I have updated the issue to avoid using meta and instead use separate `-data`, as suggested in https://github.com/json-api/json-api/issues/1097.

**Comment by richmolj** [comment]
@brainwipe I think I get where you're coming from, and for certain tagging schemes I'd agree with you. Let me try to flesh out this example, and give a few others, to address these points.

> Are tags a means to an end (filtering Posts) or are they a resource on its own? Would you be filtering Posts by tag? Does it make sense for a tag exist without a Post? It seems to me that a tag is a child entity of Post, not its own resource.

This is the key distinction. A Tag could be its own resource, and it could make sense to CRUD them independent of posts.

My company has an internal website where administrators can manage tags, so they appear when external users type into an autocomplete. I've also dealt with tags that have additional relationships - `parent` in the case of hierarchical tags, `creator` to record the administrator (not necessarily an inferred current_user) for auditing purposes. I've also added a `weight` attribute before to power a more intelligent query system.

That said, there are maybe some better examples than tagging. The main ones that usually come up are:
- Creating an order with its line items. Each line item has a `product` relationship, a `quantity` attribute, an `is_gift` attribute, etc. Your basic Amazon shopping cart.
- A sign-up form where the user enters their name and phone number (User resource) and also their address (Address resource) and maybe a few other things like credit card info in the same form.

Another scenario that I personally have is data ingestion. I provide an API and a separate team is feeding data into that API. This means hundreds of thousands of requests per day just for data ingestion. This would increase exponentially (slowing down the data ingestion process and increasing load on my server) if I weren't providing my own hand-rolled implementation of nested relationships at the moment.

I think using `attributes` would not be viable in the above scenarios.

> to be sure that this isn't huge complexity for the sake of not rethinking the server side domain model

Just to be clear, everything in this proposal is optional - nobody needs to opt-in to this additional complexity unless they want it.

**Comment by brainwipe** [comment]
> Just to be clear, everything in this proposal is optional - nobody needs to opt-in to this additional complexity unless they want it.

If we're not talking about json:api 1.1 then my anxiety is reduced but I still fear that this is opening up huge complexity; it's important for us (as developers) to attempt the simplest thing that works. Based on that...

You need to be more specific with your domain:

> A Tag could be its own resource, and it could make sense to CRUD them independent of posts.

Is not enough to answer my question. Are tags _in your example_ the point of the system. If you're taking this from your real world example, it seems that a `Tag` is used to filter [aggregates](http://martinfowler.com/bliki/DDD_Aggregate.html), or to use REST terminology _resources_. `Tag`, therefore, is not a resource in of itself, it is a child of the thing you're actually after. It's a grouping mechanism. A `tag` can have child properties but a tag is not dealt with in isolation. If you want to remove a tag from the system then you're really removing it from all the Post objects.

For your other examples:
Please let me know if I'm missing something.

And `Order` is made up from `LineItems`. Surely, the back end would create the `Order` on the first POST of a `LineItem` as it has no meaning when it is empty. The client then returns `LineItem` with the `Order` as a relationship.

For the `User` and `Address`, would you access a list of `Address` without the `User` information? What would that mean? Could you have an `Address` without a user? I think that's the very definition of a child entity.

I can't comment on your data ingestion but I imagine that there is a similar solution to it.

Specifications like json:api are supposed to guide people into creating RESTful Hypermedia services. Nesting resources goes against REST as the client is no longer dealing with a single resource at a time. Complexity is the enemy of productivity and integration.

**Comment by richmolj** [comment]
@brainwipe tags can be dealt with isolation. My administrators are dealing with them in isolation when they curate them on the internal site. Not sure what the source of confusion is here. Are you saying you think `/tags` never exists?

> Surely, the back end would create the Order on the first POST of a LineItem as it has no meaning when it is empty. The client then returns LineItem with the Order as a relationship.

The problem is all the line items must be validated at the same time. Otherwise the first POST succeeds, the second fails, and there's an invalid order/line item hanging out in my database.

I'm not sure how there would be a similar solution to the data ingestion issue. The problem is avoiding excessive network calls, your proposed solution still requires more network calls than would  otherwise be needed. Unless I am misunderstanding?

> Nesting resources goes against REST as the client is no longer dealing with a single resource at a time. 

So, would you favor removing `included` from the spec as well, since you are no longer dealing with a single resource type? I would also assume you do not favor any version of #795?

> Specifications like json:api are supposed to guide people into creating RESTful Hypermedia services. Nesting resources goes against REST as the client is no longer dealing with a single resource at a time.

Then I think we'd have to say jsonapi is failing, as by far the most requested item is dealing with more than a single resource at a time. Reading #795's description, it does seem a goal of jsonapi to support this common request.

I think this would still (ideally) be in 1.1. My point was - if you personally think this is a bad practice and do not want to deal with multiple resources in a single request, you can simply ignore all of this and nothing would change for you.

**Comment by brainwipe** [comment]
OK, I'll try to get my point across with different language and examples. I'm going to move away from the tag example, because it's too trivial to make points about Domain Driven Design. Thank you for persevering.

A resource does not need to match your domain entities. A resource can, by its nature, be composite. That is HTTP1.0. For example, an `Order` table in your database should not be the same as `Order` in your API. It is good design to disconnect these things. If you make them the same then it means any change to your database is a change to all the clients that consume your API. The API should be built from the point of view of the client, not the quickest way to develop code to give access to the database.

Given that, and `Order` from the API point of view might look like this:

```
GET
example.com/orders/1

{
  "links": {
    "self": "http://example.com/order",
  },
  "data": [{
    "type": "orders",
    "id": "1",
    "attributes": {
      "customerId": 1234,
      "products": [
          { productId: 5678; quantity: 4},
          { productId: 9876; quantity: 1}
      ]
    },
   <snip other stuff>
```

Composite resources like this hide the implementation behind the API so that the client doesn't need to know your inner workings. 

The included feature of the Spec isn't ideal, I admit that but it is a query. No state is mutated through it and you are only asking for resource _with their relations_, not any batch of resources. Whereas what you propose (and #795, which you'll see I'm also against) is performing multiple mutations in a single request. By doing so you lose a lot of power of REST because you will need to invent an error reporting system that deals with each of the operations.

What would you return in the header if the data in the Order was malformed (Bad Request 400) and a product link could not be created because it referenced a product that no longer existed (Not Found 404)? What would you return?

In the composite, you're dealing with the whole structure in one go. That is your transactional domain boundary. 

> I think this would still (ideally) be in 1.1. My point was - if you personally think this is a bad practice and do not want to deal with multiple resources in a single request, you can simply ignore all of this and nothing would change for you.

Sadly, this is not the case. A benefit of REST is that you don't have to write 100s of pages of documentation to back it up. A benefit of Hypermedia is that your API is discoverable, you don't need documentation to explain URL building. A benefit of json:api is that I can say "it's REST, it's json:api, here are the objects, here is your root url and credentials and a page of API explanation and off you go". For me to be compliant with json:api, I must implement all of it. If this (or operations) makes it into the core spec, I'll need to implement that too. That's the whole point of having a spec.

**Comment by sandstrom** [comment]
@brainwipe I think the majority of json-api implementors are writing both the server and client, making it less of an issue. Also it could be an extension (although I wouldn't mind having it in the core spec).

Even if it wouldn't be useful to you, it may be useful to others. Most of the common RFCs that underpin the Internet contain many things that I never use.

**Comment by brainwipe** [comment]
My point is that it's not about whether it's useful as a solution to a problem; my problem is that it's offering up a poor solution to a problem.

More than happy to have it in an extension as I describe here: http://discuss.jsonapi.org/t/an-appeal-to-simplicity/651

**Comment by richmolj** [comment]
@brainwipe I think we agree on most things, actually.
- We agree a resource does not need to match your database.
- We agree the inner-workings of the server should be hidden to the client.
- We agree overall on DDD and composite objects, though mileage may vary on the specifics. In particular, I tend to lean against #795 for the same reasons you do (I would use composite objects here and don't particularly feel the need for this to be part of the base spec).
- We agree hypermedia is cool 😄 
- We agree the less docs the better!

I'd also like to give an example of where I would _not_ use nested resources. Hold your breath, I'm gonna bring up tags again 😬 

``` ruby
{
  data: {
    type: 'products',
    attributes: { name: 'whirligig', tags: ['fun', 'cool'] } 
  }
}
```

See, we agree! Underneath the hood I could have a `tags` table, a polymorphic `taggings` table, what have you. But the client is agnostic to that implementation. I have no problems with the above.

Now, let's say our site works with this implementation. Six months later my product owner comes up to me and we decide we're going to add `weight` to tags to power some search functionality. Administrators will go into a `tag management` section of the site, and edit the weights. We also need to record every `weight` change for historical purposes - what the value was, what it became, and who made the change.

So an admin goes in, sees a list of tags, then clicks a specific tag to edit it. On the edit form, they see all of the historical changes made to this tag (including links to the user who made the change), and an input to change the weight themselves. When a tag is edited, the change needs to apply to universally - any products associated with this tag, 'design ideas' associated with this tag, etc, all should be affected.

So to load this form I likely hit `/api/tags/1?include=changes.user`. When changing weight, I PATCH `/api/tags/1` with a new `weight` in `attributes`. In other words, a tag is a completely independent resource, managed separately with its own endpoints.

> - We agree hypermedia is cool 😄 

So, now let's say I'm looking at a product detail page. Makes sense to hit `/api/products/1?include=tags`, right? And/or add `links` to the product's `relationships` to fetch the tags for discoverability and lazy-loading. Putting the tags in `attributes` prevents this - there's no standard for `links`, `meta` (maybe one of them is a 'primary' tag), or `relationships` (link the history). In addition, I'm telling the client 'the tag is an attribute of the product' when really it's an independent resource (a relationship). So I'd be referring to it as a resource/relationship in one place and an attribute in others, which I find counterintuitive and against the spirit of DDD/the spec itself.

Note the API still doesn't match my database (which has a `taggings` join table), and the client is still agnostic to the server's inner-workings. Since tags are a relationship, we have hypermedia links in place. But we still want to support a form where the user can either select an existing tag, or add their own (the weight will be defaulted under the hood). And we want to submit that form all-at-once, to validate everything at the same time and reduce network roundtrips. Hence, embedded relationships.

One thing I'd bring up at this point, is to keep our original API backwards-compatible we now support tags as both an attribute and a relationship. Yuck. Now, if we had _started_ with tags as a separate relationship we would not have had this problem. Which brings me to

> The API should be built from the point of view of the client, not the quickest way to develop code to give access to the database.

I agree with the latter half of this statement. However, my API serves ~10 different departments in my company, and each department has different projects that hit the API. In addition, the same API powers a website I write in Ember. I literally _can't_ build from the point of view of the client, since my API is used by so many different clients it's impossible to keep track of all their various use cases. So I tend to keep my resources separate and expose the relationships, which I believe leads to greater flexibility and avoids scenarios like the above. Logic gets pushed to the clients, so they can deal with all the objects and relationships as _they_ see fit.

> What would you return in the header if the data in the Order was malformed (Bad Request 400) and a product link could not be created because it referenced a product that no longer existed (Not Found 404)? What would you return?

I think this is more relevant to #795. This scenario is already part of the current spec, where the Order can pass product IDs in `relationships`.

> For me to be compliant with json:api, I must implement all of it.

Could you elaborate here? If you are writing a re-usable generic client, I agree with you. But if you're writing the API itself, I'm not sure where you would be forced to do this. In other words, I can currently write a fully-compliant API where I always use `links` and never use `included`. Similarly, I could still choose to use composite objects and `/relationships` endpoints instead of nested resources. So I'm not really sure what the problem is.

OK, coming in for a landing...

> A benefit of json:api is that I can say "it's REST, it's json:api, here are the objects, here is your root url and credentials and a page of API explanation and off you go".

I think this is the biggest difference in perspective we have. For me, it doesn't work this way. Almost immediately, someone asks me how to deal with nested resources (or mistakenly believes this is already part of the spec, as noted in the issue description). And I have to tell them, "well, you can read all of #795 and all of its linked issues, then read up on this library I wrote to hack our customized-version of doing this for ruby, then read up on this library I wrote for Ember to do the same. Oh, you use a Python client? You gotta go write an extension to that client and document that, too". And also write a separate documentation page specific to my API and how it deals with nested relationships. I'm actually trying to _reduce_ the amount of 'de facto' documentation a consumer has to deal with.

> More than happy to have it in an extension

Philosphically, I can agree. I think it really depends on how 'concrete' or 'official' extensions are treated and I could use some input from the maintainers here. In other words, emberjs currently has an `EmbeddedRecordsMixin` that works for most APIs, but they've specifically said it won't work (or really _try_ to work) for jsonapi because it's not part of the spec. Which means myself and many others write a bunch of hacks to roll it ourselves. Nothing can really be open-sourced in a useful way since there isn't a spec to go off of (your client library supports X way of doing this, your server supports Y way, etc).

As long as client and server libraries would be able to broadly adopt such an extension, I'd be game. If extensions are treated as 'experimental', and thus ignored by client and server libraries, then I'm still left developing a bunch of hacks and my problem is not solved. That seems to have been the case with the prior extensions.

Phew! Sorry for the wall of text. Let's try to wrap this up, since my guess is we could go back and forth forever. I am thinking of adding a summary of your perspective to the 'cons' in the issue description, and link to your comments. Then the discussion can move on to the actual implementation, but people can just downvote the issue if they agree with your points. Sound like a reasonable plan? I'd also be good to move this to a separate issue, gist, or chat, so we could continue without derailing the rest of the discussion.

**Comment by brainwipe** [comment]
Thanks for the time and effort you've put in. 

> So I'd be referring to it as a resource/relationship in one place and an attribute in others, which I find counterintuitive and against the spirit of DDD/the spec itself.

I completely agree. If you include `tag` as an aggregate then it is not an entity.

> And we want to submit that form all-at-once, to validate everything at the same time and reduce network roundtrips.

Validation should not occur across aggregates, each aggregate contains its own boundary. If you need to bind them together, you need a new aggregate... or in REST terms: resource.

> However, my API serves ~10 different departments in my company, and each department has different projects that hit the API

If you choose for your API to be generic across all possible clients then you have tightly coupled the logic of your system to the clients. I can see that some resources might be shared across multiple clients but many would not be.

> What would you return in the header if the data in the Order was malformed (Bad Request 400) and a product link could not be created because it referenced a product that no longer existed (Not Found 404)? What would you return?
> 
> I think this is more relevant to #795. This scenario is already part of the current spec, where the Order can pass product IDs in relationships.

I've had to re-read your original post for this one! This is equally relevant here. You are creating another resource which may generate a different error to your resource. What error do you report if there is a problem? Your tag resources are tiny, what if they are huge ones with 50 attributes?

> For me to be compliant with json:api, I must implement all of it.
> Could you elaborate here? If you are writing a re-usable generic client, I agree with you. But if you're writing the API itself, I'm not sure where you would be forced to do this.

I'm doing both. However, if I am writing an API and I say to an external client "we're using json:api" then you must implement it all; or you're not using it. You're using bits of it. Well, which bits? And all of a sudden you're writing documentation about how your system differentiates from the standard. There will be those cases but the more complex the spec, the more of them there will be.

> Almost immediately, someone asks me how to deal with nested resources

HTTP isn't hierarchical, it's linked. REST is built on it and in turn so is Hypermedia and json:api. If you're building hierarchical systems then you either need to add more aggregates (resources) and go composite or need a different model.

I feel we've both made our points well and respectfully disagree. I'll step back and let you get one with the specifics of your proposal. Thanks again!

**Comment by richmolj** [comment]
@brainwipe though I still fundamentally disagree, I think this was a valid and constructive conversation. I've added this to the 'cons' section of the issue, and linked to relevant comments within this thread.

For those of us who still wish to proceed, full speed ahead 🚀

**Comment by brainwipe** [comment]
I have another (new, relevant) question. What happens if the new child entity you're creating also has children that need creating? How many levels deep do you go?

**Comment by richmolj** [comment]
@brainwipe the current thought is N levels deep. I separated this out into 'Scenario C' in the proposal, so if there's contention there we can tweak/disregard it from the rest of the proposal.

**Comment by ethanresnick** [comment]
@richmolj Thanks for starting this thread! I think it's productive to have a conversation about sideposting independent of the broader operations use cases, _especially if operations is going to be a separate media type_. And @brainwipe: thanks for adding a useful and considered counterview.

I've been thinking about this for a while, and will try to respond soon.

**Comment by bfad** [comment]
One of the cons listed is "Impossible to reference an embedded resource in more than one relationship." Could this be resolved using a JSON Pointer like the error object does? It would involve second type of resource identifiers that uses "pointer" instead of "id". For example:

``` javascript
{
  data: [{
    type: 'posts',
    attributes: { title: 'my title' },
    relationships: {
      tags: {
        data: [
          { type: 'tags', attributes: { name: 'mytag' } }
        ]
      }
    }
  },
  {
    type: 'posts',
    attributes: { title: 'my title' },
    relationships: {
      tags: {
        data: [
          { pointer: '/data/0/relationships/tags/data/0', type: 'tags' }
        ]
      }
    }
  }]
}
```

Or, maybe add a top-level relationships attribute for these cases (or use "included"):

``` javascript
{
  data: [{
    type: 'posts',
    attributes: { title: 'my title' },
    relationships: {
      tags: {
        data: [
          { pointer: '/relationships/tags/data/0', type: 'tags' }
        ]
      }
    }
  },
  {
    type: 'posts',
    attributes: { title: 'my title' },
    relationships: {
      tags: {
        data: [
          { pointer: '/relationships/tags/data/0', type: 'tags' }
        ]
      }
    }
  }],
  relationships: {
    tags: {
      data: [{ type: 'tags', attributes: { name: 'mytag' } }]
    }
  }
}
```

**Comment by richmolj** [comment]
@bfad I think you're second example is close to how you'd do it with "sideposting" and temporary IDs. [See this comment from @ethanresnick linked in the body of this issue](https://github.com/json-api/json-api/issues/536#issuecomment-96635505).

If we consider this "must have" functionality I would agree with switching gears and fleshing out a sideposting proposal. That said, I'd prefer this proposal be more "dumbed down", tackling this functionality in the scope of #795. As @ethanresnick says:

> Problems of this sort only become more common when this strategy is used in conjunction with the bulk extension. For example, suppose that I want to create multiple articles at once and tag both of them with the same new tag...this is impossible. Or, on a job board, I want to add multiple job postings at once and link them with the same new employer.

So, it seems more relevant to #795. That said, if the maintainers consider something like this critical to the proposal's success, I'm down for rewriting a sideposting proposal to accommodate it.

**Comment by bfad** [comment]
@richmolj Thanks for the link to the "sideposting" solution. Personally, I would advocate that we don't need a solution to that problem for this proposal to be useful — it's an edge case that is more of a problem for bulk creation than nested creation.

**Comment by jgadrow** [comment]
@brainwipe - I also appreciate your feedback about a more RESTful way to handle this functionality though I will confess I am not certain how best to implement as a separate resource. My original proposal was to just have the clients do the work of mulitple connections (because this isn't something that would be *frequently* needed so the performance isn't so server-crushing for my use case). However, I was told by *literally* the rest of the team that this wasn't acceptable and they would rather adopt an unofficial means of doing this behavior than have to make so many requests... even as infrequent as it would be. So if you have an alternate proposal for my scenario, I'd love to hear it. :)

I think an example that is *not* just simple tags and encompasses the very core of a data set for which this would be an optimization we could allow for clients would be helpful. To that end, I have put together an example that should account for:
- Recursive processing.
- Processing multiple relationships (also recursively).
- Implying relationships if they are of a single-inheritance nature.
- A structure which is "reasonable" to assume that it could be "infinitely" nested.

To cover all of the above points, we just use a typical nested set. On the backend, this is a SQL database using the nested set pattern. So we have some additional attributes (left and right) that are *not* leaked to the client. This is because if we get write-heavy we could change to a more traditional hierarchy structure, we could change to a NoSQL data storage for which left/right aren't necessary, etc. I am pointing this out just so it is known that we are actively *avoiding* tying the request to the database. The "parent" relationship exists independent of storage mechanism and, thus, is part of the request.

Based upon the proposal, it seems like my request would look like the following:

```
{
  data: {
    attributes: {
      is_required: false
    },
    relationships: {
      children: [
		{
		  data: {
			attributes: {
			  is_required: true
			},
			relationships: {
				children: [],
				// The "parent" relationship is implied in this context.
			},
			type: 'nested-set'
		  }
		}
	  ],
      parent: {
        data: {
          attributes: {
            is_required: true
          },
          relationships: {
              children: [
                {
                  data: {
                    attributes: {
                      is_required: true
                    },
                    relationships: {
                        children: [],
                        // The "parent" relationship is implied in this context.
                    },
                    type: 'nested-set'
                  }
                }
              ],
              parent: {
                data: {
                  attributes: {
                    is_required: true
                  },
                  relationships: {
                      children: [
                        {
                          data: {
                            attributes: {
                              is_required: true
                            },
                            relationships: {
                                children: [],
                                // The "parent" relationship is implied in this context.
                            },
                            type: 'nested-set'
                          }
                        }
                      ],
                      parent: null
                  },
                  type: 'nested-set'
                }
              }
          },
          type: 'nested-set'
        }
      }
    },
    type: 'nested-set'
  }
}
```

Note: I am aware I could nest this more nicely if the "grandparent" was our intended target, but I'm not going to assume the client has their request optimized just for my benefit. Additionally, I included my "is_required" attribute just to indicate that these *do* have their own attributes that are tied to the domain logic.

I should expect a successful response that looks like this (obviously plus any links, meta-data, etc.):

```
{
  data: {
  	attributes: {
  	  is_required: false,
  	}
  	id: 'uuid-1',
  	relationships: {
  	  children: [
	    {
		  id: 'uuid-2',
		  type: 'nested-set'
		}
	  ],
	  parent: {
		id: 'uuid-3',
		type: 'nested-set'
	  }
  	},
  	type: 'nested-set'
  },
  included: [
    {
	  data: {
	    attributes: {
		  is_required: true,
		},
		id: 'uuid-2',
		relationships: {
		  children: [],
		  parent: {
		    id: 'uuid-1',
			type: 'nested-set'
		  }
		},
		type: 'nested-set'
	  }
	},
    {
	  data: {
	    attributes: {
		  is_required: true,
		},
		id: 'uuid-3',
		relationships: {
		  children: [
		    {
			  id: 'uuid-4',
			  type: 'nested-set'
			},
		    {
			  id: 'uuid-1',
			  type: 'nested-set'
			}
		  ],
		  parent: {
		    id: 'uuid-5',
			type: 'nested-set'
		  }
		},
		type: 'nested-set'
	  }
	},
    {
	  data: {
	    attributes: {
		  is_required: true,
		},
		id: 'uuid-4',
		relationships: {
		  children: [],
		  parent: {
		    id: 'uuid-3',
			type: 'nested-set'
		  }
		},
		type: 'nested-set'
	  }
	},
    {
	  data: {
	    attributes: {
		  is_required: true,
		},
		id: 'uuid-5',
		relationships: {
		  children: [
		    {
			  id: 'uuid-6',
			  type: 'nested-set'
			},
		    {
			  id: 'uuid-3',
			  type: 'nested-set'
			}
		  ],
		  parent: null
		},
		type: 'nested-set'
	  }
	},
    {
	  data: {
	    attributes: {
		  is_required: true,
		},
		id: 'uuid-6',
		relationships: {
		  children: [],
		  parent: {
		    id: 'uuid-5',
			type: 'nested-set'
		  }
		},
		type: 'nested-set'
	  }
	}
  ]
}
```

Is this correct? Have I overlooked something or made a critical error that I'm just not able to see right now?

**Comment by jimbojetlag** [comment]
It is hard to believe that this issue still exists, while the context example of creating a post with some tags is a classic app of early days of web. How is this problem unsolved, and I don't mean just in the scope of json-api. There are tons of apps out there with this use case, what did those people do to implement them?

This might indicate that some fundamentals might not be in the right place.

**Comment by lolmaus** [comment]
They implemented custom API endpoints by hand for every individual case. :unamused: And they (we) keep doing that.

**Comment by davidbnk** [comment]
We are using @richmolj proposal and it worked very well for us so far

**Comment by jimbojetlag** [comment]
Funny, here is an open issue for wordpress api unresolved since 2014 to 'Assign Categories and Tags when creating a new Post.': https://github.com/WP-API/WP-API/issues/561

Of course @richmolj proposal is pragmatic, but then again it's nothing but RPC. I'm not opposed to it, but it puts many RESTful people at unrest. I think that's the reason that this issue does not get anywhere.

Unfortunately, REST is not a practical solution. Just looks at the popular APIs, how many of them follow the REST principles? This is just enough proof that people who wanted to get their project out of the door did not adhere to REST principles.

I'd argue that things like this issue are out of the scope of REST, hence it all boils down to 2 choices:
1. If you want to stick to REST, close this issue.
2. If you want to make something that works, go with @richmolj proposal.

**Comment by jameswestnz** [comment]
Hey team - hope you don't mind another opinion. I've been keeping an eye on this issue for a while as it's certainly something I've considered in the past, but quite simply ignored because of the possible complexity both front end and back - more so in the sense that there are possibly infinite potentials that we're trying to cover. Exploring a way to make this all work within a standard specification is a great idea in my mind.

But... I feel something of this significance needs a somewhat unanimous agreement to the implementation. Quite simply because a specification like this lives or dies because of it's loyal users AND maintainers. I don't think this can be rushed.

My opinion regarding sticking to REST principles: We **should** stick to them where we can. Where we can't, we need to use them as a measure of how far "off track" we are, allowing us to decide if we're doing the right thing.

I think what has been proposed & discussed here is a great step forward to solving this problem. However as @jimbojetlag has pointed out, this issue is much bigger than the json:api spec, and we're struggling to get consensus like others before us. We're going to step on toes either way, so I think that's something we need to accept.

I believe the current status of this proposal is a step too far (adding properties such as "updated-data"), and "off track" for my liking. I think that we can take it back or step or two, so here's what I'm thinking...

1. Allow for deeply nested **CREATEs** and **UPDATEs** (examples below). If any of these events fail, the server should roll back any changes made, and provide the related errors.
2. Complex associations such as creating two new `Article`s with a common new `Tag` should be left to transactions (I'm keen to provide some thought's here too)
3. Leave  **DELETEs** for a seperate request/transactions. This is for good reason, the previous example requests are not modifying the existence of the related resource, instead they're signifying a relation to the current resource. It's like saying: "I want to update a post, associate it to tag 3, which I am now deleting."

This then leaves us with something a lot simpler, following the standard we already have in place. No new properties. No complex sideloading or Ephemeral IDs. Examples of what I believe to be fair compromise (Sorry @richmolj, I know they don't solve all your use-cases):

### Create a new article, a new tag, and associate them
```javascript
POST /articles?include=tags HTTP/1.1
Content-Type: application/vnd.api+json
Accept: application/vnd.api+json

{
  "data": {
    "type": "articles",
    "attributes": {
      "title": "New article"
    }
    "relationships": {
      "tags": {
        "data": [
          { "type": "tags", "attributes": { "name": "The new tag" } }
        ]
      }
    }
  }
}
```
The result of the directly above request would result in 1 tag for this article.

### Update an article and replace all tags with the tags provided. This may include tags yet to be created
```javascript
PUT /articles/1?include=tags HTTP/1.1
Content-Type: application/vnd.api+json
Accept: application/vnd.api+json

{
  "data": {
    "id": 1,
    "type": "articles",
    "attributes": {
      "title": "Existing article"
    }
    "relationships": {
      "tags": {
        "data": [
          { "type": "tags", "id": 2 },
          { "type": "tags", "attributes": { "name": "Another new tag" } }
        ]
      }
    }
  }
}
```
The result of the directly above request would result in 2 tags for this article.

### Update an article, only adding or updating the passed tags
```javascript
PATCH /articles/1?include=tags HTTP/1.1
Content-Type: application/vnd.api+json
Accept: application/vnd.api+json

{
  "data": {
    "id": 1,
    "type": "articles",
    "attributes": {
      "title": "Existing article"
    }
    "relationships": {
      "tags": {
        "data": [
          { "type": "tags", "id": 2, "attributes": { "name": "Updated tag 2's name" } },
          { "type": "tags", "attributes": { "name": "Yet another new tag" } }
        ]
      }
    }
  }
}
```
The result of the directly above request would result in 3 tags for this article.

### And a somewhat unplanned bonus: create a new tag, and two new articles, and associate them
```javascript
POST /tags?include=articles HTTP/1.1
Content-Type: application/vnd.api+json
Accept: application/vnd.api+json

{
  "data": {
    "type": "tags",
    "attributes": {
      "title": "Our fancy new tag"
    }
    "relationships": {
      "articles": {
        "data": [
          { "type": "articles", "attributes": { "name": "I have a tag" } },
          { "type": "articles", "attributes": { "name": "And so do I" } }
        ]
      }
    }
  }
}
```

I know I'm likely missing something... hit me up!

**Edit:** Reading over the comments again, it feels like what I have posted was the original intention, but the wants & needs evolved throughout the comments. Maybe the above examples with the touch of simplification helps us get to the next step?

**Comment by ezekg** [comment]
I'm also a long-time lurker here, so I hope you don't mind me chiming in as well.

@jameswestnz I think that's the best solution to the main issue here, and the solution I have been using in production for awhile now in an API-based product. It's clean and simple, and a lot of popular frameworks like Rails already support nested transactions like that.

I don't think introducing a new syntax/properties like the OP suggests is ideal, when it could be simplified by not allowing deletes as you suggested (which like you said, is odd to begin with in this context). As it is, the spec is relatively minimalistic (and I love that), and adding in additional properties like what is in the OP only convulutes things and makes implementation harder.

(Question though related to your examples, @jameswestnz, are the `included` query params really needed? Your requests don't actually contain included data, so I thought that was odd.)

**Comment by jameswestnz** [comment]
> (Question though related to your examples, @jameswestnz, are the included query params really needed? Your requests don't actually contain included data, so I thought that was odd.)

Great question @ezekg - I wondered if I should put that in, it was just a concept I was toying with.... But the logic there was let the server know that there are "embedded" resources to be processed. Now that it's been pointed out, I'm not convinced it's needed either.

From a specification point of view, I think it's safe to assume that if "embedded" resources are present that they should be processed. This being the case, those GET parameters aren't required.

**Comment by richmolj** [comment]
After creating this issue 8 months ago, it looks like we've fallen into the same trap as the [many](https://github.com/json-api/json-api/issues/202) [other](https://github.com/json-api/json-api/issues/205) [issues](https://github.com/json-api/json-api/issues/753) that were [created](https://github.com/json-api/json-api/issues/536) in the [past](https://github.com/json-api/json-api/issues/795). I'd once again argue that the ***specific*** solution matters vastly less than having ***a*** solution.

I believe JSONAPI should be a GraphQL competitor. This is one key area preventing such adoption. Instead, we struggle with indecision while GraphQL gains more and more market share, making the entire conversation moot.

If you're concerned about REST, I'd note our main competitor (GraphQL) is not RESTful. If you're concerned about verbosity/complexity of the payload, I'd note GraphQL has tooling and clients to mitigate the issue.

Not having a spec that accomodates relationships prevents library authors from adding functionality and improving the overall ecosystem. My guess is the vast majority of JSONAPI developers use a client to interact with the server (you don't want to parse `included` and `relationships` yourself, right?). Well, guess what - those client developers would love to add this functionality, but are prevented from doing so because the spec has not addressed this. The sooner this is in the spec itself, the sooner we can all forget about it - you're going to use a client and will not deal with the payload directly, anyway.

I'm one of those library authors ([jsonapi_suite](https://github.com/jsonapi-suite/jsonapi_suite)). Since submitting the original issue, I've moved my libraries to "sideposting", which is more intuitive and accomodates a wider variety of scenarios. If we focus on the "transferring state" **premise** of REST - rather than getting bogged down in the details of a specific REST implementation - it's RESTful as well.

Let's take the current spec, fetching a post and sideloading tags:

```ruby
{
  data: {
    type: "posts",
    id: "1",
    attributes: { title: "Man Bites Dog" }
    relationships: {
      tags: {
        data: [
          { id: "1", type: "tags" }
        ]
      }
    }
  },
  included: [
    {
      id: "1",
      type: "tags",
      attributes: { name: "Important" }
    }
  ]
}
```

Now let's say we want to update the post title and tag name in a single request. Simple `PUT` the exact same payload (with different attributes) to the corresponding endpoint.

This makes sense from a REST point of view. Our "state" is this big hash, and we are simply moving the hash back and forth (transferring state).

Of course, whenever we transfer state to the server we need a corresponding HTTP verb. But in this case, we have multiple resources which requires multiple verbs. So accomodate this, I added `method` to the resource identifier:

```ruby
{
  data: {
    type: "posts",
    id: "1",
    attributes: { title: "Man Bites Dog" }
    relationships: {
      tags: {
        data: [
          { id: "1", type: "tags", method: "destroy" }
        ]
      }
    }
  }
}
```

There's no need to put anything in `included` in this example, though I think it should still be valid.

OK so let's recap what we have so far: mirror the same payload and send it to the server, adding a verb to each resource. We're transferring state and passing a verb just like always, but now it works for N levels of nested resources.

The only thing this doesn't handle is `create`. These resources would not have an `id`, which means we wouldn't be able to assemble a sideposting payload. In addition, the client needs a way to match its in-memory resource with the one created by the server (in order to assign `id`).

So we will add `temp-id`, a random uuid sent in the request that comes back on the corresponding resource in the response.

```ruby
{
  data: {
    type: "posts",
    id: "1",
    attributes: { title: "Man Bites Dog" }
    relationships: {
      tags: {
        data: [
          { temp-id: "s0m3uu1d", type: "tags", method: "create" }
        ]
      }
    }
  },
  included: [
    {
      type: "tags",
      temp-id: "s0m3uu1d",
      attributes: { name: "Important" }
  ]
}
```

We now have full support for relationships: fetching them (via [?include](http://jsonapi.org/format/#fetching-includes)), querying them (my proposal is [here](https://github.com/json-api/json-api/issues/1175)), and creating/updating/deleting them (via "sideposting"). All we did was mirror the existing spec, adding a verb for each resource and a `temp-id` on create. We've taken a major leap forward in becoming a GraphQL competitor, and at least honored the *principles* of RESTful design.

I'm also happy to say that APIs written in this way do not require relationship endpoints. In my experience this is the trickiest, least-understood part of the spec, and avoiding them has only made my life easier.

And we no longer need to deal with payload complexity, since library authors can abstract this away (just like they do with sideloading).  For example, here are the above scenarios using my [jsorm](https://jsonapi-suite.github.io/jsorm/) library:

```js
// create
tag =  new Tag({ name: 'Important' });
post.tags.push(tag);
post.save({ with: 'tags' }).then(() => { ... });

// update
tag.name = 'Updated';
post.save({ with: 'tags' });

//destroy
tag.isMarkedForDestruction(true);
post.save({ with: 'tags' });
```

I even have a [sample application](https://gist.github.com/richmolj/c7f1adca75f614bb71b27f259ff3c37a#sample-application) with a nested form, which addresses all the scenarios in the originating issue. I currently use this for both nested web forms and speedy ETL processes successfully.

If we'd like changes, let's make 'em! I'm not tied to the naming, and my main concern is just to get **something** approved so I can write libraries around it. But let's not punt on commonly-seen problems that affect a large number of developers...problems that our main competitor **is** addressing. This issue is absolutely solveable - let's pick a solution and move on. I've recently talked to @dgeb about this, so maybe we can get some input from the core team on this as well to focus the conversation.

**Comment by ethanresnick** [comment]
I've been off this repo for a while and had to take a break as an editor (I got bogged down with other work), but I recently started a new API project so I wanted to revisit JSON:API. I agree 100% with @richmolj's core contention that, at this point, we just need something. It doesn't need to be perfect, but it needs to exist. Until we get that, I'm personally using what I described [here](https://github.com/json-api/json-api/issues/795#issuecomment-312362533) because it works and it's spec-compliant today.

Really, though, I think the state of this issue points to a deeper problem: the core team, myself included, just doesn't have enough bandwidth to address even the major issues that remain to be sorted. That includes not only this issue, but also (imo) #383, #1175, #862, #600, #684, etc. That's why, when I was last active in JSON:API, I was pushing so hard to get extensions done: extensions will unblock the community on all these things. **We need to get out of the way so that user land can figure this stuff out**, while we offer a way for their solutions to be backed by a spec (the spec of the extension).

To that end, I have some more time now, and I would be happy to jump back in to push extensions across the finish line, I'm just not sure what the next step is or how far away we still are on that. If @dgeb and @wycats / @steveklabnik could provide some thoughts on the way forward, that would be much appreciated. @tkellen, I'd of course love your input too, but my sense is that you've also had to move on to other things.

**Comment by antonkomarev** [comment]
Guys, thank you for your effort and such a big research job done so far! Special thanks goes to @richmolj who compiled all the stuff together and to @ethanresnick for the idea of embedding new entities.

After reading both of the issues (#795, #1089) I've understood that there is no win-win solution for all of the use cases. I've tried to summarize all the stuff in my head, add my own vision of the situation and try to propose another one solution.

### Good old PUT method

> The PUT method is already defined to overwrite a resource with a complete new body, and cannot be reused to do partial changes.

By following REST specification we can use PUT method to delete\create\update relationships.

#### Preface

Imagine we have `Article` with `id: 1` which has 3 relations: `author`, `tags`, `changes`.

`GET /articles/1`
```json
{
  "data": {
    "type": "article",
    "id": "1",
    "attributes": {
      "title": "Test title",
      "body": "Test body",
    },
    "relationships": {
      "author": {
        "data": {"type": "user", "id": "678"}
      },
      "tags": {
        "data": [
          {"type": "tag", "id": "123"},
          {"type": "tag", "id": "124"}
        ]
      },
      "changes": {
        "data": [
          {"type": "change", "id": "98"}
        ]
      }
    }
  },
  "included": [
    {
      "type": "user",
      "id": "678",
      "attributes": {
        "name": "Test user",
      }
    },
    {
      "type": "tag",
      "id": "123",
      "attributes": {
        "name": "JSON API Examples",
        "slug": "json-api-examples"
      }
    },
    {
      "type": "tag",
      "id": "124",
      "attributes": {
        "name": "Brand new one",
        "slug": "brand-new-one"
      }
    }
    {
      "type": "change",
      "id": "98",
      "attributes": {
        "title": "Very old title",
        "body": "Very old body"
      }
    }
  ]
}
```

#### Test case

So if we will pass empty `relationships.tags.data` array all the tags relationships will be dropped:

`PUT /articles/1`
```json
{
  "data": {
    "type": "article",
    "id": "1",
    "attributes": {
      "title": "Updated title",
      "body": "Updated body",
    },
    "relationships": {
      "tags": {
        "data": []
      }
    }
  }
}
```

#### Analysis

This solution has it's own side-effects. We must send all the relationships in the request body, otherwise it will be evaluated as dropping all the relations which aren't in the list. Because PUT replacing all the body of the resource.

Imagine that we have `author` relationship and forgot to add it to the list of relationships... `author` relationship will be dropped then.
This will lead to an increase amount of the transferred data. Because before we start to update any entity, we must collect complete list of it's relations and then post them back.

#### Possible solution

We could replace body (attributes) of the main resource as RFC tells us, but replace only relations which present in `relationships` object. If we will bypass empty data objects\arrays for `author` and `tags` relationships - both of them will be dropped and `changes` will be unaffected. Is this REST enough?

> **NOTE:** This replace action shouldn't delete related resources, it's only relationships deletion. For related resource deletion only DELETE method could be used with appropriate endpoint for it.

#### Advanced test case

In this test we are:
1. deleting one tag (because it's omitted from `relationships.tags.data` array;
2. adding a new tag (it shouldn't be stored in `included` array because tag entity not yet exists, thats why I've placed it in custom `embedded` array) which discoverable with `pointer` what will help us to keep strict consistency that each resource should be represented in request\response only once.
3. deleting all the `changes` relationships.

`PUT /articles/1`
```json
{
  "data": {
    "type": "article",
    "id": "1",
    "attributes": {
      "title": "Updated title",
      "body": "Updated body",
    },
    "relationships": {
      "tags": {
        "data": [
          {"type": "tag", "id": "123"},
          {"pointer": "/embedded/0"}
        ]
      },
      "changes": {
        "data": []
      }
    }
  },
  "embedded": [
    {
      "type": "tag",
      "attributes": {
        "name": "Fresh tag",
        "slug": "fresh-tag"
      }
    }
  ],
  "included": [
    {
      "type": "tag",
      "id": "123",
      "attributes": {
        "name": "JSON API Examples",
        "slug": "json-api-examples"
      }
    }
  ]
}
```

Will be glad to hear criticism of this solution.

**Comment by ethanresnick** [comment]
@a-komarev JSON:API [already allows](http://jsonapi.org/format/#crud-updating-relationships) replacing a relationship with an empty array using the PATCH method:

```
{
  "data": {
    "type": "article",
    "id": "1",
    "relationships": {
      "tags": {
        "data": []
      }
    }
  }
}
```

Because of that, I don't think we get anything extra by supporting PUT. Basically, anything PUT can do PATCH can do too, but with less data transfer overhead and less need for optimistic locking.

Regarding the solution you propose for the more complex cases (creating + linking a related resource at the same time), I think it's interesting, especially because it shows how we might extend "sideposting" to update requests like PUT/PATCH. That said, the solution you're proposing is very similar to the solutions discussed before... at this point, I think most of the solutions have been explored, and we really just need to implement something rather than discuss new options.

**Comment by ethanresnick** [comment]
@richmolj Would you be up to write a PR speccing out a solution to this that could (finally) be included in the base spec? I would enthusiastically back a solution along the lines of sideposting, esp. if it handled bulk creation at the same time (a la my proposal [here](https://github.com/json-api/json-api/issues/795#issuecomment-132955057)). 

I'm less comfortable backing a 'nested resource objects' approach, because I think we really should solve bulk creation sooner or later and, in that context, I don't like the limitations of nested resource objects (i.e., not being able to use the same to-be-created resource in multiple relationships). Also, I think it's confusing for the contents of a relationship to sometimes be a pointer to a resource object (linkage) and sometimes be a full resource object. 

That said, if you're not comfortable with a sideposting approach, you could instead spec out the nested resource objects approach as an extension. Doing that would, it seems, be a good stress test for the latest set of proposed extension rules (#1195), and having that extension would be amazing for a lot of people.

Three notes if you go route of specifying sideposting for the base spec:

1. Allowing `data` to take an array of resource objects when POSTing to a collection endpoint seems so obvious that I suspect it would be an uncontroversial change. Especially since we already allow data to be an array (of linkage) when POSTing to a relationship endpoint. That said, we'd have to make it an optional feature (as 1.0 servers that don't support an array would return 400). We'd also have to tweak our language around creation a bit in the array case. For example, we'd probably temper our recommendation around the `Location` response header and allow the POSTing of resources that don't directly. Also, we might want to mint a specific URL for this "array of resources not supported" error, which 1.1 servers that don't want to support bulk creation could use in the newly-proposed `type` field on error objects (see #1195).

2. While I called the collection of sideposted resources `"embedded"` in #795, I don't remember the rationale for this and `"included"` might actually make more sense. 1.0 servers that don't support `included` in a POST would also return 400.

3. Sideposting does involve introducing a new pointer type (whether a JSON Pointer or a temp id) that can go in a relationship object during the POST. Supporting this would be part of the opt-in semantics. So, the spec would say something to the effect that "If an `included` key is present in a POST, relationship objects in the request may have this new type of pointer."

**Comment by richmolj** [comment]
@ethanresnick I'll be working with @beauby on a Sideposting PR coming in the next week or two. I've already switched my libraries to this approach ([explained here](https://jsonapi-suite.github.io/jsonapi_suite/concepts#sideposting)) so that's preferred by me as well. Appreciate your thoughts and will incorporate into the PR.

**Comment by ethanresnick** [comment]
@richmolj Awesome! I'm super, super excited to keep this rolling and see what y'all come up with.

**Comment by ezekg** [comment]
I'm not a fan of the additional `method` key or the temp IDs—just seems odd and it makes everything seem more complicated than it should be e.g. in Rails, AR knows that when the associated ID isn't specified in the create, then it should be a new resource; if it does exist, it should be associated; if the association is set to an empty array i.e. `[]` or `nil`, then it should be deleted.

I'd assume other ORMs/frameworks handle things in a like-manner, and there's no reason that the API can't implement something like that, which is so much simpler than reimplementing HTTP methods within the JSON payload. (Or maybe I'm overlooking something important here?)

And besides, like @jameswestnz mentioned earlier, saying  "I want to update a post, associate it to tag 3, which I am now deleting" is just odd and is a major smell imo. I'd really, really like to keep things simple, but I understand if it's not the direction you all want to go in.

I was originally attracted to this spec because of its simplicity, but the above discussions are honestly a little much for me and not at all the direction I wanted the convo to go in. [I've been using a simplified version of embedded relationships for an API product of mine](https://keygen.sh/docs/api/#accounts-create) that works well with Rails with minimal effort (keep in mind that that API may not be 100% spec compliant).

But anyways, enough bike-shedding for me.

**Comment by rpkilby** [comment]
The side-posting draft has been posted as PR #1197.

**Comment by dgeb** [comment]
@richmolj has this been fully superceded by #1197? Should this be closed in favor of moving discussion there?

**Comment by richmolj** [comment]
@dgeb I'd say so

**Comment by sandstrom** [comment]
@ezekg Small note on your comment above: the temporary IDs are used by the client-side, they don't exist for the benefit of the server-side (e.g. Rails), although they need to cycled through the backend.

Some of these things are a bit more complicated than they seem, it may seem like needless complication, but some of these additions would be very useful 😄 

*(this has been the 3rd most upvoted issue for a reason)*
